src/metpy/calc/thermo.py

```diff
@@ -4205,6 +4206,223 @@ def k_index(pressure, temperature, dewpoint, vertical_dim=0):
+
+    # Interpolate to find temperature and mixing ratio at 950, 850, 700, and 500 hPa
+    (
+        (t950, t850, t700, t500),
+        (r950, r850, r700, r500),
+        (th950, th850, th700, th500)
+    ) = interpolate_1d(
+        units.Quantity([950, 850, 700, 500], 'hPa'),
+        pressure, temperature_k, mixing_ratio, potential_temp_k,
+        axis=vertical_dim,
+    )
+
+    th_a = th950
+    r_a = r950
+
+    th_b = 0.5 * (th850 + th700)
+    r_b = 0.5 * (r850 + r700)
+
+    th_c = th500
+    r_c = r500
+
+    alpha = -10 * units.kelvin  # Empirical adjustment
+
+    # Latent heat of vaporization of water - different from MetPy constant
+    #  Using different value, from paper, since GDI unlikely to be used to
+    #  derive other metrics
+    l_0 = 2.69E6 * (units.joule / units.kilogram)
+
+    # Temperature math from here on requires kelvin units
+    eptp_a = th_a * np.exp((l_0 * r_a) / (Cp_d * t850))
+    eptp_b = th_b * np.exp((l_0 * r_b) / (Cp_d * t850)) + alpha
+    eptp_c = th_c * np.exp((l_0 * r_c) / (Cp_d * t850)) + alpha
+
+    if t950.size == 1:
+        is_array = False
+    else:
+        is_array = True
+
+    # Calculate C.B.I.
+    beta = 303 * units.kelvin  # Empirical adjustment
+    l_e = eptp_a - beta  # Low-troposphere EPT
+    m_e = eptp_c - beta  # Mid-troposphere EPT
+
+    # Gamma unit - likely a typo from the paper, should be units of K^(-2) to
+    #  result in dimensionless CBI
+    gamma = 6.5e-2 * (1 / units.kelvin)
+    zero_kelvin = 0 * units.kelvin
+
+    if is_array:
+        # Replace conditional in paper for array compatibility.
+        # Will set CBI for any l_e < 0 to 0
+        l_e[l_e <= zero_kelvin] = zero_kelvin
+    else:
+        l_e = max(l_e, zero_kelvin)
+    column_buoyancy_index = gamma * (l_e * m_e)
+    # Convert to magnitude and dimensionless to avoid unit issue from typo
+    column_buoyancy_index = column_buoyancy_index.magnitude
+
+    # Calculate Mid-tropospheric Warming Index
+    tau = 263.15 * units.kelvin  # Threshold
+    mu = -7 * (1 / units.kelvin)  # Empirical adjustment
+
+    t_diff = t500 - tau
+    if is_array:
+        t_diff[t_diff <= zero_kelvin] = zero_kelvin
+    else:
+        t_diff = max(t_diff, zero_kelvin)
+    mid_tropospheric_warming_index = mu * t_diff
+    mid_tropospheric_warming_index = mid_tropospheric_warming_index.magnitude
+
+    # Calculate Inversion Index
+    sigma = 1.5 * (1 / units.kelvin)  # Empirical scaling constant
+    s = t950 - t700
+    d = eptp_b - eptp_a
+
+    inv_sum = s + d
+    if is_array:
+        inv_sum[inv_sum >= zero_kelvin] = zero_kelvin
+    else:
+        inv_sum = min(inv_sum, zero_kelvin)
+
+    inversion_index = sigma * inv_sum
+    inversion_index = inversion_index.magnitude
+
+    # Calculate Terrain Correction
+    p_3 = 18
+    p_2 = 9000 * units.hectopascal
+    p_1 = 500 * units.hectopascal
+    p_sfc = pressure[0]
+    terrain_correction = p_3 - (p_2 / (p_sfc - p_1))
+
+    # Convert all to 'dimensionless'
```

Comment:
This section should be able to be removed because each element will already have the dimensionless units since we are not taking the magnitude of each portion with the removal of those lines above.